tests/zfs-tests/tests/functional/userquota/userspace_encrypted.ksh

```diff
@@ -0,0 +1,80 @@
+#!/bin/ksh -p
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright 2019, loli10K <ezomori.nozomu@gmail.com>. All rights reserved.
+#
+
+. $STF_SUITE/include/libtest.shlib
+. $STF_SUITE/tests/functional/userquota/userquota_common.kshlib
+
+#
+# DESCRIPTION:
+# 'zfs userspace' and 'zfs groupspace' can be used on encrypted datasets
+#
+#
+# STRATEGY:
+# 1. Create both un-encrypted and encrypted datasets
+# 2. Receive un-encrypted dataset in encrypted hierarchy
+# 3. Verify encrypted datasets support 'zfs userspace' and 'zfs groupspace'
+#
+
+function cleanup
+{
+	destroy_pool $POOLNAME
+	rm -f $FILEDEV
+}
+
+log_onexit cleanup
+
+FILEDEV="$TEST_BASE_DIR/userspace_encrypted"
+POOLNAME="testpool$$"
+typeset -a POOL_OPTS=('' # all pool features enabled
+    '-d' # all pool features disabled
+    '-d -o feature@userobj_accounting=enabled' # only userobj_accounting enabled
+    '-d -o feature@project_quota=enabled') # only project_quota enabled
+DATASET_ENCROOT="$POOLNAME/encroot"
+DATASET_SENDFS="$POOLNAME/sendfs"
+
+log_assert "'zfs user/groupspace' should work on encrypted datasets"
+
+for opts in "${POOL_OPTS[@]}"; do
+	# Setup
+	truncate -s $SPA_MINDEVSIZE $FILEDEV
+	log_must zpool create $opts -o feature@encryption=enabled $POOLNAME \
+		$FILEDEV
+
+	# 1. Create both un-encrypted and encrypted datasets
+	log_must zfs create $DATASET_SENDFS
+	log_must eval "echo 'password' | zfs create -o encryption=on" \
+		"-o keyformat=passphrase -o keylocation=prompt " \
+		"$DATASET_ENCROOT"
+	log_must zfs create $DATASET_ENCROOT/fs
+
+	# 2. Receive un-encrypted dataset in encrypted hierarchy
+	log_must zfs snap $DATASET_SENDFS@snap
+	log_must eval "zfs send $DATASET_SENDFS@snap | zfs recv " \
+		"$DATASET_ENCROOT/recvfs"
+
+	# 3. Verify encrypted datasets support 'zfs userspace' and 'zfs groupspace'
+	log_must eval "zfs userspace $DATASET_ENCROOT/fs > /dev/null"
+	log_must eval "zfs groupspace $DATASET_ENCROOT/fs > /dev/null"
+	log_must_retry "unsupported" 3 \
+		eval "zfs userspace $DATASET_ENCROOT/recvfs > /dev/null"
+	log_must_retry "unsupported" 3 \
+		eval "zfs groupspace $DATASET_ENCROOT/recvfs > /dev/null"
```

Comment:
Was the issue here that it might return unsupported until it completes?  I wasn't able to reproduce this locally, and it's not what I'd expect. Unless you're able to reproduce this I'd suggest we simplify this like so:

```diff
@@ -65,13 +65,12 @@ for opts in "${POOL_OPTS[@]}"; do
        log_must eval "zfs send $DATASET_SENDFS@snap | zfs recv " \
                "$DATASET_ENCROOT/recvfs"
 
-       # 3. Verify encrypted datasets support 'zfs userspace' and 'zfs groupsp
-       log_must eval "zfs userspace $DATASET_ENCROOT/fs > /dev/null"
-       log_must eval "zfs groupspace $DATASET_ENCROOT/fs > /dev/null"
-       log_must_retry "unsupported" 3 \
-               eval "zfs userspace $DATASET_ENCROOT/recvfs > /dev/null"
-       log_must_retry "unsupported" 3 \
-               eval "zfs groupspace $DATASET_ENCROOT/recvfs > /dev/null"
+       # 3. Verify encrypted datasets support 'zfs userspace' and
+       # 'zfs groupspace'
+       log_must zfs userspace $DATASET_ENCROOT/fs
+       log_must zfs groupspace $DATASET_ENCROOT/fs
+       log_must zfs userspace $DATASET_ENCROOT/recvfs
+       log_must zfs groupspace $DATASET_ENCROOT/recvfs
 
        # Cleanup
        cleanup
```

Reply:
My local builder seems to be a lot slower than those running on build.zfsonlinux.org; i can sometimes triggers failures if i remove `log_must_retry()`

```
--- Configuration ---
Runfiles:        /usr/share/zfs/runfiles/zfs-tests.run
STF_TOOLS:       /usr/share/zfs/test-runner
STF_SUITE:       /usr/share/zfs/zfs-tests
STF_PATH:        /var/tmp/constrained_path.QRLQ
FILEDIR:         /var/tmp
FILES:           /var/tmp/file-vdev0 /var/tmp/file-vdev1 /var/tmp/file-vdev2
LOOPBACKS:       /dev/loop0 /dev/loop1 /dev/loop2
DISKS:           loop0 loop1 loop2
NUM_DISKS:       3
FILESIZE:        4G
ITERATIONS:      1
TAGS:            functional
STACK_TRACER:    no
Keep pool(s):    rpool
Missing util(s): mmap_libaio pamtester umask wait

/usr/share/zfs/test-runner/bin/test-runner.py  -c "/usr/share/zfs/runfiles/zfs-tests.run" -T "functional" -i "/usr/share/zfs/zfs-tests" -I "1"
Test: /usr/share/zfs/zfs-tests/tests/functional/userquota/setup (run as root) [00:01] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/userquota/userspace_encrypted (run as root) [00:07] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/userquota/userspace_encrypted (run as root) [00:07] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/userquota/userspace_encrypted (run as root) [00:03] [FAIL]
Test: /usr/share/zfs/zfs-tests/tests/functional/userquota/userspace_encrypted (run as root) [00:07] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/userquota/userspace_encrypted (run as root) [00:06] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/userquota/userspace_encrypted (run as root) [00:03] [FAIL]
Test: /usr/share/zfs/zfs-tests/tests/functional/userquota/cleanup (run as root) [00:01] [PASS]

Results Summary
FAIL	   2
PASS	   6

Running Time:	00:00:38
Percent passed:	75.0%
Log directory:	/var/tmp/test_results/20201113T161214

Tests with results other than PASS that are expected:

Tests with result of PASS that are unexpected:

Tests with results other than PASS that are unexpected:
    FAIL userquota/userspace_encrypted (expected PASS)
```

```
Test: /usr/share/zfs/zfs-tests/tests/functional/userquota/userspace_encrypted (run as root) [00:03] [FAIL]
16:12:48.10 ASSERTION: 'zfs user/groupspace' should work on encrypted datasets
16:12:48.20 SUCCESS: zpool create -o feature@encryption=enabled testpool14821 /var/tmp/userspace_encrypted
16:12:48.23 SUCCESS: zfs create testpool14821/sendfs
16:12:49.35 SUCCESS: eval echo 'password' | zfs create -o encryption=on -o keyformat=passphrase -o keylocation=prompt  testpool14821/encroot
16:12:49.45 SUCCESS: zfs create testpool14821/encroot/fs
16:12:49.52 SUCCESS: zfs snap testpool14821/sendfs@snap
16:12:49.65 SUCCESS: eval zfs send testpool14821/sendfs@snap | zfs recv  testpool14821/encroot/recvfs
16:12:49.65 TYPE        NAME   USED  QUOTA  OBJUSED  OBJQUOTA
16:12:49.65 POSIX User  root  1.50K   none        1      none
16:12:49.66 SUCCESS: zfs userspace testpool14821/encroot/fs
16:12:49.67 TYPE         NAME   USED  QUOTA  OBJUSED  OBJQUOTA
16:12:49.67 POSIX Group  root  1.50K   none        1      none
16:12:49.68 SUCCESS: zfs groupspace testpool14821/encroot/fs
16:12:49.69 TYPE        NAME   USED  QUOTA  OBJUSED  OBJQUOTA
16:12:49.69 POSIX User  root  1.50K   none        1      none
16:12:49.69 SUCCESS: zfs userspace testpool14821/encroot/recvfs
16:12:49.70 TYPE         NAME   USED  QUOTA  OBJUSED  OBJQUOTA
16:12:49.70 POSIX Group  root  1.50K   none        1      none
16:12:49.71 SUCCESS: zfs groupspace testpool14821/encroot/recvfs
16:12:49.95 SUCCESS: zpool destroy -f testpool14821
16:12:50.10 SUCCESS: zpool create -d -o feature@encryption=enabled testpool14821 /var/tmp/userspace_encrypted
16:12:50.15 SUCCESS: zfs create testpool14821/sendfs
16:12:51.13 SUCCESS: eval echo 'password' | zfs create -o encryption=on -o keyformat=passphrase -o keylocation=prompt  testpool14821/encroot
16:12:51.19 SUCCESS: zfs create testpool14821/encroot/fs
16:12:51.22 SUCCESS: zfs snap testpool14821/sendfs@snap
16:12:51.32 SUCCESS: eval zfs send testpool14821/sendfs@snap | zfs recv  testpool14821/encroot/recvfs
16:12:51.34 TYPE        NAME   USED  QUOTA  OBJUSED  OBJQUOTA
16:12:51.34 POSIX User  root  1.50K   none        -         -
16:12:51.34 SUCCESS: zfs userspace testpool14821/encroot/fs
16:12:51.35 TYPE         NAME   USED  QUOTA  OBJUSED  OBJQUOTA
16:12:51.35 POSIX Group  root  1.50K   none        -         -
16:12:51.36 SUCCESS: zfs groupspace testpool14821/encroot/fs
16:12:51.38 cannot get used/quota for testpool14821/encroot/recvfs: unsupported version or feature
16:12:51.38 ERROR: zfs userspace testpool14821/encroot/recvfs exited 255
```

Keeping `log_must_retry()` should prevent spurious failures.

Reply:
OK.  Well in that case we should still add something like `(( $? != 0 )) && log_fail` after each `log_must_retry` call.  Otherwise it won't fail properly after 3 retries and will incorrectly pass, see the `log_must_busy` helper function as an example.  How about something like this:

```ksh
function log_must_unsupported
{
        log_must_retry "unsupported" 3 "$@"
        (( $? != 0 )) && log_fail
}

       ....

       # 3. Verify encrypted datasets support 'zfs userspace' and
       # 'zfs groupspace'
       log_must zfs userspace $DATASET_ENCROOT/fs
       log_must zfs groupspace $DATASET_ENCROOT/fs
       log_must_unsupported zfs userspace $DATASET_ENCROOT/recvfs
       log_must_unsupported zfs groupspace $DATASET_ENCROOT/recvfs
```

Reply:
With `log_must_unsupported()` it is indeed failing if i inject failures with
```c
stap -g -e 'probe module("zfs").function("zfs_ioc_userspace_many").return { $return = 95; }'
```
so it is safe to assume it will catch real issues: it wasn't doing that with just `log_must_retry()`, thank you for catching this.